db/hints/sync_point_service.hh

```diff
@@ -0,0 +1,302 @@
+
+    public:
+        /// \brief Creates a task.
+        ///
+        /// The task isn't running yet; use task::run() in order to run it.
+        task(task_map& parent, sstring typ, sync_point_id id);
+
+        /// \brief Runs the task with the given function.
+        ///
+        /// This function should only be called once in task's lifetime.
+        /// The function is passed an abort source which will be triggered
+        /// when the task is cancelled - when it happens, the function should
+        /// abandon its work and complete ASAP, possibly with
+        /// abort_requested_exception.
+        void run(noncopyable_function<future<> (abort_source&)>&& f);
+
+        /// \brief Checks the result of the task.
+        ///
+        /// \returns true if the task completed successfully,
+        ///          false if it is still running.
+        /// \throws an exception if the internal function threw an exception.
+        bool check_result();
+
+        /// \brief Waits until the task completes, with a specified timeout.
+        ///
+        /// \returns future which resolves after the task stops running, containing
+        ///          its result (which can be an exception).
+        future<> wait_for_result(clock_type::time_point timeout);
+
+        /// \brief Cancels this task now.
+        ///
+        /// Triggers task's abort source. The task will asynchronously wait
+        /// until it stops running and then will remove itself from the map.
+        ///
+        /// It is safe to call this function multiple times; only the first
+        /// cancellation has an effect.
+        void cancel();
+
+        /// \brief Cancels this task now and waits until it finishes.
+        ///
+        /// Triggers task's abort source and waits synchronously until
+        /// the task stops running and removes itself from the map.
+        ///
+        /// \returns A future which resolves after the task removes itself
+        ///          from the parent map and is destroyed.
+        future<> cancel_and_finish();
+
+        /// \brief (Re)sets internal timer, scheduling cancellation of the task.
+        ///
+        /// By default, the internal timer is not running. This function
+        /// can be used to implement TTL for the task.
+        ///
+        /// Calling this function for the second time resets the timeout.
+        /// If the task was cancelled, it has no effect.
+        void schedule_expiry(clock_type::duration timeout);
+    };
+
+private:
+    /// \brief Collection of coordinator-side tasks.
+    ///
+    /// A coordinator-side task creates follower-sync points and periodically
+    /// queries the followers in order to check if hints were already replayed,
+    /// or an error has occurred.
+    ///
+    /// Coordinator tasks are scheduled for expiry after the internal function stops
+    /// (see `completed_coordinator_task_ttl`).
+    ///
+    /// Tasks should be created on task 0 only.
```

Comment:
shard 0?

Maybe id % smp::count to distribute things.

Reply:
We are enforcing a limit on the number of tasks on a node, so this needs to be tracked on a single shard.